tiledb/tests/test_group.py

```diff
@@ -636,3 +689,76 @@ def test_consolidation_and_vac_no_config(self):
         tiledb.Group.vacuum_metadata(path)
 
         assert len(vfs.ls(meta_path)) == 1
+
+    def test_string_metadata(self, capfd):
+        # this test ensures that string metadata is correctly stored and
+        # retrieved from the metadata store. It also tests that the metadata
+        # dump method works correctly for string metadata.
+        uri = self.path("test_ascii_metadata")
+        tiledb.Group.create(uri)
+
+        grp = tiledb.Group(uri, "w")
+        grp.meta["abc"] = "xyz"
+        grp.close()
+
+        grp = tiledb.Group(uri, "r")
+        assert grp.meta["abc"] == "xyz"
+        grp.meta.dump()
+        assert_captured(capfd, "Type: DataType.STRING_UTF8")
+        grp.close()
+
+    def test_array_or_list_of_strings_metadata_error(self):
```

Comment:
Actually, was wondering how this handles empty strings like

`np.array(["1", "2", "3", "", "f", "o", "o"], dtype="U1")` (in the middle)
`np.array(["1", "2", "3", "f", "o", "o", ""], dtype="U1")` (trailing)
`np.array([""], dtype="U1")` (by itself)

Reply:
Thanks for your comment! It's a good one!

The way this works is by assigning a byte to each string, even if it's empty. ASCII ("S") strings are a bit easier to visualize. So, when you assign the following array as metadata: `np.array(["1", "2", "3", "", "f", "o", "o"], dtype="S1")`, what you're essentially sending to the `Group::put_metadata` C++ API is `b'123\x00foo'`. As you can see, there is a null byte (`\x00`) which represents the empty string.

Similarly, `np.array([""], dtype="S1")` sends `b'\x00'`.

For `np.array(["1", "2", "3", "", "f", "o", "o"], dtype="U1")`, we send `b'1\x00\x00\x002\x00\x00\x003**\x00\x00\x00\x00**\x00\x00\x00f\x00\x00\x00o\x00\x00\x00o\x00\x00\x00'`. 

Here, null bytes are also used as padding because the characters can be represented in ASCII. The empty string is represented by null bytes, which you can see between the `**` symbols.

I hope this answers your question.

Reply:
Awesome - thanks for the explaination!